### ISSUE TITLE
How to also get train Map metrics/scores and how to visualize inference on images/video

### ISSUE THREAD

**Issue by johnk2hawaii** [ArgoHA/custom_d_fine#10]
Excellent repo btw. Note that I had to downgrade torch to 2.5.1 to get it to run in Colab as of the date of this comment.

I am looking for three things. 

**First Question**: how do I get metrics for the train set? I understand that map is usually only calculated on the validation set but is there a way to access any loss metrics for the training set (so as to inspect for overfitting)

**Second question** I have questions about inference. I was wondering what the built in method is to visualize and save inferenced images. Right now I followed your advice and I have: 

```
from src.infer.torch_model import Torch_model
import torch
import cv2
import matplotlib.pyplot as plt

model = Torch_model(
        model_name="s",
        model_path="/content/output/models/baseline2_2025-05-15/model.pt",
        n_outputs=7,
        input_width=640,
        input_height=640,
        conf_thresh=0.5,
        rect=False,
        half=False,
        keep_ratio=False,
)

img = cv2.imread('/content/custom_dataset/images/IMG_2282_jpeg_jpg.rf.510f3bc14c3e0aa378b192199d01cae6.jpg') # read 1 image in a folder
res = model(img)

```
For now I have wriiten a custom block of code that seems to work (but I would like to use what ever elegant built in method exists):

```
class_labels = ['fish', 'jellyfish', 'penguin', 'puffin', 'shark', 'starfish', 'stingray']
def visualize_predictions(img, res, class_labels):
    """Draws bounding boxes on the image."""
    for i in range(len(res[0]['scores'])):
        x, y, w, h = res[0]['boxes'][i]
        conf = res[0]['scores'][i]
        class_label = res[0]['labels'][i]
        x1, y1, x2, y2 = int(x - w/2), int(y - h/2), int(x + w/2), int(y + h/2)
        cv2.rectangle(img, (x1, y1), (x2, y2), (0, 255, 0), 2)
        cv2.putText(img, f"{class_labels[int(class_label)]} {conf:.2f}",
                    (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

    plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
    plt.axis('off')
    plt.show()

```

**Last question**, how do we do inference and save results on a video? . I am looking to have the bounding boxes overlaying the original feed.

Cheers and thanks

**Comment by ArgoHA** [comment]
Hey, thanks for your feedback!

1) metrics are not calculated on the train set. What you can do - customize training loop and compute metrics not only on val_loader, but train_loader too. I never do that because it is slow and for me don't give a good insight, that's why it is not implemented.

2) You can use `python -m src.dl.infer`, it saves visualized predicted boxes. If you want that in your code, check out [this function](https://github.com/ArgoHA/custom_d_fine/blob/main/src/dl/infer.py#L13). What I think about now, maybe a method under the Torch_model that plots bboxes would be easier to use, what do you think? Smth like `visualized_image = model.plot(img, preds)`

3) Video inference is not implemented yet, but it's a great idea, thanks, I totally forgot about it. I think I will update infer script, so it works on folder with images or videos. Output will be the same video but with bboxes plotted on the frames. I'll let you know when it is live.

**Comment by ArgoHA** [comment]
Yes, if you pass args to the training script and not change the config file, then you also need to do that in other scripts too. Updated the readme.

Other things noted, will get back soon

**Comment by pra-dan** [comment]
I was able to infer on video and save results as frames (just as they happen right now) with a function change and modified parsing logic.

**Comment by ArgoHA** [comment]
Hey everyone. Pushed an update, now you can run inference on a folder of videos or a folder with images. Output is pretty much the same.

**Comment by johnk2hawaii** [comment]
Thanks so much. Looking forward to seeing your instance segmentation integration! I will close this issue now
